pkg/lib/catalogsource/catalogsource_update.go

```diff
@@ -0,0 +1,173 @@
+package catalogsource
+
+import (
+	"context"
+	"reflect"
+	"sync"
+
+	"github.com/operator-framework/api/pkg/operators/v1alpha1"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/api/client/clientset/versioned"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// mu is a package scoped mutex for synchronizing catalog source updates
+var mu sync.Mutex
+
+/* UpdateStatus can be used to safely update the status of the provided catalog source. Note that the
+status values are updated to the values from catsrc in their entirety when using this function.
+
+• logger: used to log errors only
+
+• client: used to fetch / update catalog source status
+
+• catsrc: the CatalogSource to use as a source for status updates. Callers are
+responsible for updating the catalog source status values as necessary.
+*/
+func UpdateStatus(logger *logrus.Entry, client versioned.Interface, catsrc *v1alpha1.CatalogSource) error {
+	mu.Lock()
```

Comment:
TLDR; Using the locks protects the CatalogSource "shared resource" from two threads updating status at same time.

We've got two separate controllers attempting to update the same CatalogSource objects. If it was the same controller you could guarantee how it was being updated. Right now the status updates don't interfere with each other... the status conditions are used by the new controller, and the old controller updates its existing fields. However in the future if the existing controller updates the status conditions array, updates to this array need to be coordinated to ensure that the content and order of the array are maintained properly. Ultimately this is trying to minimize the chances of the two controllers interfering with each other.

Reply:
> Ultimately this is trying to minimize the chances of the two controllers interfering with each other.

It does that, but it also bottlenecks the workers that are processing queues of catalogsources: they're working in parallel but because this mutex is in the package scope, updates to catalogsource status have to happen serially.

The way this is generally dealt with in the other controllers is:
 - use `patch` instead of `update` whenever possible (in this case, patch should always succeed since no other controller owns that condition)
  - if the patch/update fails for any reason, the key for the catalog is put back on the queue and the change is recalculated / retried
  
 This model has the additional benefit of decoupling the controllers so that they can run in different pods if needed

Reply:
Bottom line... the way things are written they do return errors if `UpdateStatus(...)` or `Update(...)` fail, and my last change ensure that the errors are propagated to the "sync" return value, so the retry should kick in as you've described. 

So it seems like I just need to remove the mutex since you're saying its not really needed.

Reply:
By the way, I added the mutex since someone was concerned about two controllers updating status.

Reply:
I did put in place a comment on the `UpdateStatus` function that says the following (only showing relevant portions here):

> ... Note that the status values are updated to the values from catsrc in their entirety when using this function.
...
• catsrc: the CatalogSource to use as a source for status updates. Callers are
responsible for updating the catalog source status values as necessary

The purpose of the documentation was to make it super clear that callers are responsible for making valid status updates when using this function. To that end I had analyzed all of the places where catalog source status is updated and found that the only offender was a section of code that blew up the entire status section:

```go
// The sourceType is valid, clear (all) status if there's existing invalid spec reason
if out.Status.Reason == v1alpha1.CatalogSourceSpecInvalidError {
    out.Status = v1alpha1.CatalogSourceStatus{}
}
```

I changed this code to this to preserve the conditions prior to resetting the other status:

```go
// The sourceType is valid, clear all status (other than status conditions array) if there's existing invalid spec reason
if out.Status.Reason == v1alpha1.CatalogSourceSpecInvalidError {
    out.Status = v1alpha1.CatalogSourceStatus{
        Conditions: out.Status.Conditions,
    }
}
```
This is the link to the diff shown above: https://github.com/operator-framework/operator-lifecycle-manager/pull/2281/files#diff-a1760d9b7ac1e93734eea675d8d8938c96a50e995434b163c6f77c91bace9990R606-R608

This is the other reason I created separate functions for updating the status... to make status updates as restricted in scope as possible.

Reply:
I think https://github.com/operator-framework/operator-lifecycle-manager/pull/2281/files#r679474664 responds, in part, to this as well.

further, it seems like this helper is still stomping the entire status, and not just a specific condition.

Reply:
fixed